Documentation/design-docs/native-hosting-assembly-load.md

```diff
@@ -0,0 +1,89 @@
+
+
+## Existing support
+[COM Activation](COM-activation.md) allows native apps to effectively load managed components (assemblies), but it requires the use of COM activation APIs and general COM related registration and setup.
+
+
+## High-level proposal
+In .NET Core 3.0 the hosting layer (see [here](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md)) ships with several hosts. These are binaries which act as the entry points to the .NET Core hosting/runtime:
+* The "muxer" (`dotnet.exe`)
+* The `apphost` (`.exe` which is part of the app)
+* The `comhost` (`.dll` which is part of the app and acts as COM server)
+* The `ijwhost` (`.dll` consumed via `.lib` used by IJW assemblies)
+
+Every one of these hosts serve different scenario and expose different APIs. The one thing they have in common is that their main purpose is to find the right `hostfxr`, load it and call into it to execute the desired scenario. For the most part all these hosts are basically just wrappers around functionality provided by `hostfxr`.
+
+The proposal is to add a new host library `nethost` which can be used by native apps to easily host managed components.
+
+*Technical note: All strings in the proposed APIs are using the `char_t` in this document for simplicity. In real implementation they are of the type `pal::char_t`. In particular:*
+* *On Windows - they are `WCHAR *` using `UTF16` encoding*
+* *On Linux/macOS - they are `char *` using `UTF8` encoding*
+
+
+## New host binary for component hosting
+Add new library `nethost` which will act as the easy to use host for loading managed components.
+The library would be a dynamically loaded library (`.dll`, `.so`, `.dylib`). For ease of use there would be a header file for C++ apps as well as `.lib`/`.a` for easy linking.
+Apps using the component hosting functionality would ship this library as part of the app. Unlike the `apphost`, `comhost` and `ijwhost`, the `nethost` will not be directly supported by the .NET Core SDK since it's target usage is not from .NET Core apps.
+
+The exact delivery mechanism is TBD (pending investigation), but it probably should include NuGet (for C++ projects) and plain `.zip` (for any consumer). The binary itself should be signed by Microsoft as there will be no support for modifying the binary as part of custom application build (unlike `apphost`).
+
+### Load managed component and get a function pointer
+``` C++
+int nethost_load_assembly_method(
+        const char_t * assembly_path,
+        const char_t * type_name,
+        const char_t * method_name,
+        const void * reserved,
+        void ** delegate);
+)
+```
+This API will
+* Locate the assembly using the `assembly_path` and its `.runtimeconfig.json` and determine the frameworks it requires to run. (Note that only framework dependent components will be supported for now).
+* If the process doesn't have CoreCLR loaded (more specifically `hostpolicy` library)
```

Comment:
What happens if there is no runtimeconfig.json and the runtime is not loaded?  I think that may be a fairly common scenario (eg. a loosely built piece of managed code embedded into a piece of native code).

Reply:
Just like in `comhost` it's not a supported scenario right now. Currently the lack of `.runtimeconfig.json` indicates self-contained application. It would make sense to apply the same logic for components. Unfortunately supporting self-contained components is very tricky (if not impossible).

That said, the alternative would be that the host somehow magically picks a runtime. But it's not just about the runtime, it's about all frameworks.
.NET Framework didn't have this problem since the host picked the runtime to use and with it came all the frameworks (from GAC). There were no side-by-side frameworks or runtimes or anything like that.

In the .NET Core world, we need something to make the choice. So far both the runtime and the host try to stay clear of policy decisions (with various levels of success).

I assume that just like for COM activation, the components enabled for loading through this API will require to be built for it. Basically exactly because of this problem.

What we may allow is for components to load without `.runtimeconfig.json` if there already is runtime in the process. It would mean we state that we won't ever support self-contained components (since the lack of `.runtimeconfig.json` is the indication of self-contained), but that might be an OK solution.

Reply:
What if we used a slightly different pattern...
Initialize
Start
LoadAssembly
CreateDelegate
Stop

Any number of these could be collapsed (eg. Initialize implies Start, Load & Create are collapsed, etc.).  I feel that the most predominant native hosting scenario would like more control of initialization (eg. not relying on an on disk artifact).

You may then say, that this is being considered as a separate PR (per our offline discussion) and that this is just one building block of the full flow.  In which case, perhaps we add an InitializeViaRuntimeConfig (which takes a path)?

Reply:
Aside from all of the added complexity of allowing full control (discussed elsewhere)… this is basically a discussion between which part of the system should be able to participate in framework/runtime decisions.

The current proposal favors the components - basically we try to guarantee that they will work (by making sure that their framework requirements are met, and thus requiring them to specify such requirements).

If we go with the other approach where the host is favored, it would mean a much larger burden on the host to "do the right thing" for components to work.